lib/src/package_config_impl.dart

```diff
@@ -0,0 +1,178 @@
+      : version = _validateVersion(version),
+        _packages = _validatePackages(packages);
+
+  SimplePackageConfig._(int version, Iterable<SimplePackage> packages)
+      : version = _validateVersion(version),
+        _packages = {for (var package in packages) package.name: package};
+
+  /// Creates empty configuration.
+  ///
+  /// The empty configuration can be used in cases where no configuration is
+  /// found, but code expects a non-null configuration.
+  const SimplePackageConfig.empty()
+      : version = 1,
+        _packages = const <String, Package>{};
+
+  static int _validateVersion(int version) {
+    RangeError.checkValueInInterval(version, 1, _maxVersion, "version");
+    return version;
+  }
+
+  static Map<String, Package> _validatePackages(Iterable<Package> packages) {
+    Map<String, Package> result = {};
+    for (var package in packages) {
+      if (package is! _SimplePackage) {
+        // _SimplePackage validates these properties.
+        try {
+          _validatePackageData(package.name, package.root,
+              package.packageUriRoot, package.languageVersion);
+        } catch (e) {
+          throw ArgumentError.value(
+              packages, "packages", "Package ${package.name}: ${e.message}");
+        }
+      }
+      var name = package.name;
+      if (result.containsKey(name)) {
+        throw ArgumentError.value(name, "packages", "Duplicate package name");
+      }
+      result[name] = package;
+    }
+
+    // Check that no root URI is a prefix of another.
+    if (result.length > 1) {
+      // Uris cache their toString, so this is not as bad as it looks.
+      var rootUris = [...result.values]
+        ..sort((a, b) => a.root.toString().compareTo(b.root.toString()));
+      var prev = rootUris[0];
+      var prevRoot = prev.root.toString();
+      for (int i = 1; i < rootUris.length; i++) {
+        var next = rootUris[i];
+        var nextRoot = next.root.toString();
+        // If one string is a prefix of another,
+        // the former sorts just before the latter.
+        if (nextRoot.startsWith(prevRoot)) {
+          throw ArgumentError.value(
+              packages,
+              "packages",
+              "Package ${next.name} root overlaps "
+                  "package ${prev.name} root.\n"
+                  "${prev.name} root: $prevRoot\n"
+                  "${next.name} root: $nextRoot\n");
+        }
+        prev = next;
+      }
+    }
+    return result;
+  }
+
+  Iterable<Package> get packages => _packages.values;
+
+  Package /*?*/ operator [](String packageName) => _packages[packageName];
+
+  /// Provides the associated package for a specific [file] (or directory).
+  ///
+  /// Returns a [Package] which contains the [file]'s path.
+  /// That is, the [Package.rootUri] directory is a parent directory
+  /// of the [file]'s location.
+  /// Returns `null` if the file does not belong to any package.
+  Package /*?*/ packageOf(Uri file) {
```

Comment:
This logic will also have to support package roots which are a prefix of another.

The typical strategy here is to take the longest matching prefix - but it does require iterating all packages.

Reply:
Again, I was planning to *not* allow that.

Reply:
we can consolidate discussion in the other thread, resolving this one